distribution/sc-proxy-dex/src/wrapped_farm_token_merge.rs

```diff
@@ -0,0 +1,317 @@
+            require!(
+                &elem.token_id == wrapped_farm_token_id,
+                "Not a Wrapped Farm Token"
+            );
+        }
+        Ok(())
+    }
+
+    fn merge_locked_asset_tokens_from_wrapped_farm(
+        &self,
+        tokens: &[WrappedFarmToken<Self::BigUint>],
+    ) -> GenericTokenAmountPair<Self::BigUint> {
+        let locked_asset_factory_addr = self.locked_asset_factory_address().get();
+
+        if tokens.len() == 1 {
+            let token = tokens[0].clone();
+            let locked_token_amount = self.rule_of_three(
+                &token.token_amount.amount,
+                &token.attributes.farm_token_amount,
+                &token.attributes.farming_token_amount,
+            );
+
+            return GenericTokenAmountPair {
+                token_id: self.locked_asset_token_id().get(),
+                token_nonce: token.attributes.farming_token_nonce,
+                amount: locked_token_amount,
+            };
+        }
+
+        let locked_asset_token = self.locked_asset_token_id().get();
+        for entry in tokens.iter() {
+            let locked_token_amount = self.rule_of_three(
+                &entry.token_amount.amount,
+                &entry.attributes.farm_token_amount,
+                &entry.attributes.farming_token_amount,
+            );
+
+            self.locked_asset_factory_proxy(locked_asset_factory_addr.clone())
+                .deposit_tokens(
+                    locked_asset_token.clone(),
+                    entry.attributes.farming_token_nonce,
+                    locked_token_amount,
+                )
+                .execute_on_dest_context();
+        }
+
+        self.locked_asset_factory_proxy(locked_asset_factory_addr)
+            .merge_locked_asset_tokens(OptionalArg::Some(BoxedBytes::from(ACCEPT_PAY_FUNC_NAME)))
+            .execute_on_dest_context_custom_range(|_, after| (after - 1, after))
+    }
+
+    fn merge_farm_tokens(
+        &self,
+        farm_contract: &Address,
+        tokens: &[WrappedFarmToken<Self::BigUint>],
+    ) -> GenericTokenAmountPair<Self::BigUint> {
+        if tokens.len() == 1 {
+            let token = tokens[0].clone();
+
+            return GenericTokenAmountPair {
+                token_id: token.attributes.farm_token_id,
+                token_nonce: token.attributes.farm_token_nonce,
+                amount: token.token_amount.amount,
+            };
+        }
+
+        for entry in tokens.iter() {
+            self.farm_contract_merge_proxy(farm_contract.clone())
+                .deposit_tokens(
+                    entry.attributes.farm_token_id.clone(),
+                    entry.attributes.farm_token_nonce,
+                    entry.token_amount.amount.clone(),
+                )
+                .execute_on_dest_context();
+        }
+
+        self.farm_contract_merge_proxy(farm_contract.clone())
+            .merge_farm_tokens(OptionalArg::Some(BoxedBytes::from(ACCEPT_PAY_FUNC_NAME)))
```

Comment:
The reason is that we've added a check on sender part for errors. The contracts are in the same shard and if the send fails, the execution should be terminated on the sender part and the whole transaction will be reverted. When we introduced this validate functions, we didn't really know what to use and this check was the strongest and should have worked with any method. This is the reason of validate functions. They do no harm since it's just a check and I don't want to delete them because of this. Also do not want to add another function for the gas. We'll like to make as little changes as possible because we are time restricted.

Reply:
Resolved.

BTW, it was slightly easier to review the code with those checks, because sometimes one could see locally that certain properties hold (e.g. that the received amounts use a specific token). However, if those validations were not needed, it would not have been worth it to keep them just for that.